packages/loopring_v3/contracts/amm/libamm/AmmBlockReceiver.sol

```diff
@@ -18,50 +18,43 @@ library AmmBlockReceiver
     using AmmExitProcess    for AmmData.State;
     using AmmJoinProcess    for AmmData.State;
     using AmmPoolToken      for AmmData.State;
-    using AmmUpdateProcess  for AmmData.State;
-    using BlockReader       for ExchangeData.Block;
+    using AmmUpdateProcess  for AmmData.Context;
+    using BlockReader       for bytes;
 
     function beforeBlockSubmission(
-        AmmData.State      storage   S,
-        ExchangeData.Block memory   _block,
-        bytes              calldata data,
-        uint                        txIdx,
-        uint                        numTxs
+        AmmData.State      storage  S,
+        bytes              memory   txsData,
+        bytes              calldata data
         )
         internal
     {
-        AmmData.Context memory ctx = _getContext(S, txIdx);
+        AmmData.Context memory ctx = _getContext(S);
 
-        BlockReader.BlockHeader memory header = _block.readHeader();
-        require(header.exchange == address(ctx.exchange), "INVALID_EXCHANGE");
+        ctx.approveAmmUpdates(txsData);
 
-        S.approveAmmUpdates(ctx, _block);
-
-        _processPoolTx(S, ctx, _block, data);
+        _processPoolTx(S, ctx, txsData, data);
 
         // Update state
         S._totalSupply = ctx.totalSupply;
 
         // Make sure we have consumed exactly the expected number of transactions
-        require(numTxs == (ctx.txIdx - txIdx), "INVALID_NUM_TXS");
+        require(txsData.length/ExchangeData.TX_DATA_AVAILABILITY_SIZE == ctx.txIdx, "INVALID_NUM_TXS");
```

Comment:
shall we check `lengh === txIdx * TX_DATA_AVAILABILITY_SIZE`. 7/3==2, but 7!= 3*2